### ISSUE TITLE
Torch training?

### ISSUE THREAD

**Issue by tnies** [Computational-Biology-Aachen/MxlPy#40]
When we use batched torch_surrogate training (``` train_torch```), it looks like only one random batch per epoch is used. But I think there is also the variant where the data set is split into batches then the network is trained using all batches per epoch. After this the training set is shuffled and new batches are created that are used in the new epoch. It should be something like thies

```python
def _train_batched_test(
    aprox: nn.Module,
    features: pd.DataFrame,
    targets: pd.DataFrame,
    epochs: int,
    optimizer: Adam,
    device: torch.device,
    batch_size: int,
    loss_fn: Callable,
) -> pd.Series:

    x = torch.tensor(features.to_numpy().astype(np.float32), dtype=torch.float32)
    y = torch.tensor(targets.to_numpy().astype(np.float32), dtype=torch.float32)
    dataset = TensorDataset(x, y)
    train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True)
    losses = {}
    for i in tqdm.trange(epochs):
        running_train_loss = 0.0
        for xb, yb in train_loader:
            optimizer.zero_grad()
            pred = aprox(xb)
            loss = loss_fn(pred, yb)  # torch.mean(torch.abs(pred - Y))
            loss.backward()
            optimizer.step()
            running_train_loss += loss.item() * xb.size(0)
        losses[i] = running_train_loss / len(train_loader.dataset)
    return pd.Series(losses, dtype=float)

```
But I am not sure about the loss calculation over the total epoch

**Comment by marvinvanaalst** [comment]
Ah true. Weird. In the `npe._train_batched` function we do it properly. 

Which begs the question why they are two functions in the first place (apart from the way they were historically implemented). 

Do you know how the `DataLoader` is implemented? If I set `batch_size=None`, will it train the entire dataset at once? Because then we can probably also unify the `_train_full` and `_train_batched` methods

**Comment by tnies** [comment]
Docs says that default is ```batch_size=1```. A quick test showed me that when ```batch_size=None``` it seems to fall back to 1.

**Comment by marvinvanaalst** [comment]
Ah ok, but then we can emulate by just using `len(features)`.

Can you double check [the latest commit to dev](https://github.com/Computational-Biology-Aachen/MxlPy/tree/1411e644d278a9136fc36bb04421c472b5a29e78)? I unified all four functions into one and put it in `mxlpy.nn._torch.train`

Test seem to run fine and example notebook as well.

**Comment by tnies** [comment]
Seems, alright for me, but at some point we should discuss if this is right way to calculate the loss over an whole epoch. Tests with the wong surrogate resulted in the expected loss

![Image](https://github.com/user-attachments/assets/51c3f21d-028f-41fe-8197-900210c1c2f1)

I think it can be pushed to main

**Comment by marvinvanaalst** [comment]
Why not discuss it directly? What epoch loss do you have in mind?

**Comment by marvinvanaalst** [comment]
Implemented in [v0.18](https://github.com/Computational-Biology-Aachen/MxlPy/releases/tag/v0.18.0) 🥳
